packages/slice-machine/src/features/autoSave/useAutoSave.tsx

```diff
@@ -0,0 +1,114 @@
+import { useCallback, useState, useEffect } from "react";
+import { uniqueId } from "lodash";
+import { toast } from "react-toastify";
+import { Box, Button, Text } from "@prismicio/editor-ui";
+
+type UseAutoSaveArgs = {
+  debounceDelay?: number;
+  errorMessage?: string;
+  retryDelay?: number;
+  retryMessage?: string;
+};
+
+type AutoSaveState = {
+  pendingSave: (() => Promise<unknown>) | null;
+  nextSave: (() => Promise<unknown>) | null;
+};
+
+export type AutoSaveStatusType = "saving" | "saved" | "error";
+
+export const useAutoSave = (args: UseAutoSaveArgs = {}) => {
+  const {
+    debounceDelay = 500,
+    errorMessage = "An error happened while saving",
+    retryDelay = 2000,
+    retryMessage = "Retry",
+  } = args;
+  const [autoSaveState, setAutoSaveState] = useState<AutoSaveState>({
+    pendingSave: null,
+    nextSave: null,
+  });
+  const [autoSaveStatusInternal, setAutoSaveStatusInternal] =
+    useState<AutoSaveStatusType>("saved");
+  const [autoSaveStatus, setAutoSaveStatus] = useState<AutoSaveStatusType>(
+    autoSaveStatusInternal,
+  );
+
+  const setNextSave = useCallback((saveAction: () => Promise<unknown>) => {
+    setAutoSaveState((prevState) => ({
+      ...prevState,
+      nextSave: saveAction,
+    }));
+  }, []);
+
+  const executeSave = useCallback(async () => {
+    if (autoSaveState.pendingSave) {
+      setAutoSaveStatusInternal("saving");
+
+      try {
+        await autoSaveState.pendingSave();
+
+        setAutoSaveStatusInternal("saved");
+        setAutoSaveState((prevState) => ({
```

Comment:
Hum good thinking. I will try to see if it's possible and test it. It may simplify a bit the useEffect bellow even maybe.

But I'm not sure. I will try and come back to you 🙂

Reply:
Yes! 

So I fixed the bug of the retry, and I really took time to check that one.

When a save fail, we want to trigger a retry with the latest data we have (saw with Benjamin), meaning we don't want to just reply the pending. We may want to retry with the next, but if there is no next, meaning the user did not do any action in the meantime, we will retry with the pending.
Because of this it's important in case of a success to remove the pending, if not the useEffect will detect that there is a pending operation and will try to do it again, creating an endless loop.

I added comments for the tricky parts, btw.